rfcs/1111-first-class-virtual-device.md

```diff
@@ -0,0 +1,120 @@
+- Feature Name: Add virtual device information as a first class field on Relay expressions.
+- Start Date: 2021-11-20
+- RFC PR: [apache/tvm-rfcs#1111](https://github.com/apache/tvm-rfcs/pull/1111)
+- GitHub Issue: [apache/tvm#1111](https://github.com/apache/tvm/issues/1111)
+
+# Summary
+[summary]: #summary
+
+I propose adding a new field to Relay expressions, virtual_device_. This field will contain virtual device information [currently called SEScope].
+
+# Motivation
+[motivation]: #motivation
+
+ Currently, the virtual device information (called SEScope today, but we will rename it soon) is stored in on_device Relay ops. This op is a wrapper op that contains the virtual device information for an expression. This pattern method was used to encode device information in Relay before @mbs-octoml's recent work on device planning. In his work, he continued to use on_device to encode the virtual device information.
+
+At the time, we considered adding virtual_device_ as a first-class field in Relay expressions. However, we decided not to purse that avenue because not all fields of expressions are preserved during passes, so it would be a lot of work ensure that a new field is propagated correctly. Additionally, we weren't sure how fundamental the virtual devices would be.
+
+Now, we believe that virtual devices are very fundamental. Pretty much every expression in Relay needs to have one to enable heterogeneous execution. Using on_device to store the virtual device information is clunky. Normal visitors that don't care about virtual devices need to peek inside on_device ops. Additionally, to change the virtual device of an expression, we have to backtrack to the last on_device and then rewrite it. First class virtual devices eliminate a lot of this complexity, and will allow us to implement more features in device and memory planning in the future.
```

Comment:
clunky -> has proven to be fragile

Instead of talking about backtracking,  mention requires DeviceAware* visitors needed to answer the basic question 'what device should this sub-expression be evaled on'